api/api_request_logger.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2024 The VeChainThor developers
+
+// Distributed under the GNU Lesser General Public License v3.0 software license, see the accompanying
+// file LICENSE or <https://www.gnu.org/licenses/lgpl-3.0.html>
+
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+	"time"
+
+	log "github.com/inconshreveable/log15"
+)
+
+// RequestLogger logs requests to an output
+type RequestLogger struct {
+	enabled      bool
+	writerChan   chan entry
+	stopChan     chan bool
+	outputWriter func(msg string, ctx ...interface{})
+}
+
+// NewThorRequestLogger applies thor business logic
+// when returning an instance of *RequestLogger
+func NewThorRequestLogger(enabledFlag bool, enableLog bool) *RequestLogger {
+	if enabledFlag {
+		return NewRequestLogger(enabledFlag, log.Info)
+	}
+	if enableLog {
+		return NewRequestLogger(enableLog, log.Debug)
+	}
+	return NewRequestLogger(false, nil)
+}
+
+func NewRequestLogger(enabled bool, logger func(msg string, ctx ...interface{})) *RequestLogger {
+	return &RequestLogger{
+		enabled:      enabled,
+		outputWriter: logger,
+		writerChan:   make(chan entry, 100_000),
+		stopChan:     make(chan bool),
+	}
+}
+
+func (l *RequestLogger) Enabled() bool {
+	return l.enabled
+}
+
+// Handle returns an http handler to ensure requests are syphoned into the writer
+func (l *RequestLogger) Handle(handler http.Handler) http.Handler {
+	fn := func(w http.ResponseWriter, r *http.Request) {
+		// Read and log the body (note: this can only be done once)
+		// Ensure you don't disrupt the request body for handlers that need to read it
+		var bodyBytes []byte
+		var err error
+		if r.Body != nil {
+			bodyBytes, err = io.ReadAll(r.Body)
+			if err != nil {
+				log.Warn("unexpected body read error", "err", err)
+				return // don't pass bad request to the next handler
+			}
+			r.Body = io.NopCloser(io.Reader(bytes.NewReader(bodyBytes)))
+		}
+
+		l.writerChan <- entry{
+			Timestamp: time.Now(),
+			URI:       r.URL.String(),
+			Method:    r.Method,
+			Body:      string(bodyBytes),
+		}
+
+		// call the original http.Handler we're wrapping
+		handler.ServeHTTP(w, r)
+	}
+
+	// start the writer
+	l.start()
+
+	// http.HandlerFunc wraps a function so that it
+	// implements http.Handler interface
+	return http.HandlerFunc(fn)
+}
+
+func (l *RequestLogger) Stop() {
+	if !l.enabled {
+		return
+	}
+
+	// make sure any pending message is written
+	l.stopChan <- true
+	<-l.stopChan
+}
+
+func (l *RequestLogger) start() {
+	go func() {
+		for {
+			select {
+			case item := <-l.writerChan:
+				marshal, err := json.Marshal(item)
+				if err != nil {
+					log.Warn("unable to marshal api request entry", "err", err)
+					continue
+				}
+
+				l.outputWriter("[request] " + string(marshal))
```

Comment:
`log  = log15.New("pkg", "api")` might be good for filtering and can this be done via the traditional formatted way as below? 
```
log.Debug("URI", item.URI, "Method", item.Method)
```

Reply:
Yes, that was my first go-to, but I was looking for something prepended vs appended so that it's a fixed structure text instead of a variable one.

Ex. `INFO[05-17|15:23:46] [request] {"timestamp":` its easy to filter on `[request]`
Ex. `INFO[05-17|15:23:46] {"timestamp":.......WEIRD THINGS} type=request` it's not easy to filter on `[request]`

Reply:
Perhaps if it's easy to prepend the tags? It seem'd convoluted to me.